contrib/jackc/pgx.v5/pgx_tracer.go

```diff
@@ -175,6 +183,43 @@ func (t *pgxTracer) TraceConnectEnd(ctx context.Context, data pgx.TraceConnectEn
 	finishSpan(ctx, data.Err)
 }
 
+func (t *pgxTracer) TraceAcquireStart(ctx context.Context, pool *pgxpool.Pool, _ pgxpool.TraceAcquireStartData) context.Context {
+	if !t.cfg.traceAcquire {
+		return ctx
+	}
+	opts := t.spanOptions(pool.Config().ConnConfig, operationTypeAcquire, "")
+	_, ctx = tracer.StartSpanFromContext(ctx, "pgx.acquire", opts...)
+	return ctx
+}
+
+func (t *pgxTracer) TraceAcquireEnd(ctx context.Context, pool *pgxpool.Pool, data pgxpool.TraceAcquireEndData) {
+	if t.cfg.traceHold {
+		opts := t.spanOptions(pool.Config().ConnConfig, operationTypeHold, "")
+		_, holdCtx := tracer.StartSpanFromContext(ctx, "pgx.hold", opts...)
+		data.Conn.PgConn().CustomData()[customDataContextKey] = holdCtx
+	}
+
+	if !t.cfg.traceAcquire {
+		return
+	}
+
+	finishSpan(ctx, data.Err)
+}
+
+func (t *pgxTracer) TraceRelease(pool *pgxpool.Pool, data pgxpool.TraceReleaseData) {
+	if !t.cfg.traceHold {
+		return
+	}
+
+	holdCtx, ok := data.Conn.PgConn().CustomData()[customDataContextKey].(context.Context)
+	if !ok {
+		return
+	}
+
+	finishSpan(holdCtx, nil)
+	delete(data.Conn.PgConn().CustomData(), customDataContextKey)
```

Comment:
not sure if this needs to be thread-safe

Reply:
I'm not 100% sure either, but I could imagine it to be fine. At the very least, we should make sure the tests for this pass with `-race`.

Reply:
I think it's not covered by tests. I also think it's not thread-safe and I'd rather to make it clear in the Godocs or ensure it's thread-safe with a mutex.

Reply:
what do you think about leaving the context there? it can safely get overwritten by the next acquireEnd call